### ISSUE TITLE
How can I do Cross Compile for ARM64

### ISSUE THREAD

**Issue by Zhang8912** [eclipse-cyclonedds/cyclonedds#1366]
Hi All,
currently we will move our Examples of CycloneDDS on Platform ARM64. However, I meet problem again.

Firstly, the following compilers are installed for ARM64.
   sudo apt-get install  g++-7-aarch64-linux-gnu
   sudo apt-get install gcc-7-aarch64-linux-gnu

after that, I used following commands for compiling. But I met Errors with **idlc**. Did anyone know how to solve the problem or how to run cross compiling for ARM64? 
thanks in advance

 **pwd
 /home/felix/Templates/dds1/cyclonedds
 rm -rf build
 mkdir build
 cd build
 cmake -DCMAKE_INSTALL_PREFIX=/home/felix/Templates/dds1/cyclonedds -DCMAKE_C_COMPILER=/usr/bin/aarch64-linux-gnu-gcc-7 ..
 cmake --build .**

![image](https://user-images.githubusercontent.com/103993997/184095057-7456a11c-1889-4aea-93b2-b5044ef67fca.png)

**Comment by thijsmie** [comment]
Hi @Zhang8912,

When you cross-compile for another architecture the binaries you produce cannot run on the host platform. This includes the IDl compiler (`idlc`) which is the issue here. You can supply an `idlc` compiled for your host architecture on your `PATH` or simply turn off the only part of the build process that needs `idlc` at build time by passing `-DBUILD_DDSPERF=OFF` to your cmake invocation. See also https://github.com/eclipse-cyclonedds/cyclonedds/issues/1358.

**Comment by shizhuang-li** [comment]
Good idea, I have also encountered these problems. The specific reason for this problem can be viewed in the outermost CMakeLists.txt, where it is stated that if the current cross-compilation is detected during the compilation process, the compilation of idlc will be canceled, so after cross-compilation is No idlc executable will be generated.

After this, you need to read the idl file to generate the .c .h data file by using the idlc that was compiled under the x86 architecture before, and you also need to modify the example CMakeLists.txt. Here I take the HelloWorld example as an example:

```
project(helloword LANGUAGES C)
cmake_minimum_required(VERSION 3.10)

#set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -lsocket -l c")

if(NOT TARGET CycloneDDS::ddsc)
  # Find the CycloneDDS package.
  find_package(CycloneDDS REQUIRED)
endif()

# This is a convenience function, provided by the CycloneDDS package,
# that will supply a library target related the the given idl file.
# In short, it takes the idl file, generates the source files with
# the proper data types and compiles them into a library.

#idlc_generate(TARGET HelloWorldData_lib FILES "HelloWorldData.idl" WARNINGS no-implicit-extensibility)

# Both executables have only one related source file.
add_executable(HelloworldPublisher publisher.c HelloWorldData.c)
add_executable(HelloworldSubscriber subscriber.c HelloWorldData.c)

# Both executables need to be linked to the idl data type library and
# the ddsc API library.
# target_link_libraries(HelloworldPublisher HelloWorldData_lib CycloneDDS::ddsc)
# target_link_libraries(HelloworldSubscriber HelloWorldData_lib CycloneDDS::ddsc)

target_link_libraries(HelloworldPublisher CycloneDDS::ddsc)
target_link_libraries(HelloworldSubscriber CycloneDDS::ddsc)
```

**Comment by Zhang8912** [comment]
> Hi @Zhang8912,
> 
> When you cross-compile for another architecture the binaries you produce cannot run on the host platform. This includes the IDl compiler (`idlc`) which is the issue here. You can supply an `idlc` compiled for your host architecture on your `PATH` or simply turn off the only part of the build process that needs `idlc` at build time by passing `-DBUILD_DDSPERF=OFF` to your cmake invocation. See also #1358.

Hi, thanks for your reply. It works.

**Comment by eboasson** [comment]
Thanks for the feedback, @Zhang8912 ☺️
